raiden/transfer/log.py

```diff
@@ -0,0 +1,238 @@
+            '    id integer primary key autoincrement, data binary'
+            ')'
+        )
+        cursor.execute(
+            'CREATE TABLE IF NOT EXISTS state_snapshot (id integer primary key, data binary)'
+        )
+        self.conn.commit()
+
+    def write_transaction(self, data):
+        cursor = self.conn.cursor()
+        cursor.execute(
+            'INSERT INTO transactions(id, data) VALUES(null,?)',
+            (data,)
+        )
+        self.conn.commit()
+
+    def write_state_snapshot(self, identifier, data):
+        cursor = self.conn.cursor()
+        result = cursor.execute(
+            'SELECT * from state_snapshot',
+            (identifier, data)
+        )
+        result = result.fetchone()
+        if result is None:
+            cursor.execute(
+                'INSERT INTO state_snapshot(id, data) VALUES(?,?)',
+                (identifier, data)
+            )
+        else:
+            cursor.execute(
+                'UPDATE state_snapshot SET id=?, data=? WHERE id=?',
+                (identifier, data, result[0])
+            )
+        self.conn.commit()
+
+    def get_transaction_by_id(self, identifier):
+        cursor = self.conn.cursor()
+        result = cursor.execute(
+            'SELECT data from transactions where id=?', (identifier,)
+        )
+        result = result.fetchall()
+        if result != list():
+            assert len(result) == 1
+            result = result[0][0]
+        return result
+
+    def get_all_transactions(self):
+        cursor = self.conn.cursor()
+        result = cursor.execute(
+            'SELECT * from transactions'
+        )
+        return result.fetchall()
+
+    def read(self):
+        pass
+
+    def last_identifier(self):
+        cursor = self.conn.cursor()
+        result = cursor.execute(
+            'SELECT seq FROM sqlite_sequence WHERE name="transactions"'
+        )
+        result = result.fetchall()
+        if result != list():
+            assert len(result) == 1
+            result = result[0][0]
+        else:
+            result = 0
+        return result
+
+    def __del__(self):
+        self.conn.close()
+
+
+class TransactionLogFileBackend(TransactionLogStorageBackend):
+    """This is just an example for having a file backend. Not actually implemented"""
+
+    def __init__(self, filepath):
+        self.filepath = filepath
+        self.file = create_file_iff_not_existing(
+            self.filepath,
+            return_it=True,
+            mode='r+',
+            buffering=False
+        )
+        self._synced = False  #: True when the existing log is read to the end
+
+    def write_transaction(self, data):
+        pass
+
+    def write_state_snapshot(self, identifier, data):
+        pass
+
+    def read(self):
+        pass
+
+    def last_identifier(self):
+        pass
+
+    def __del__(self):
+        self.file.flush()
+        self.file.close()
+
+
+class TransactionLog(object):
```

Comment:
IMO this class can have it's interface simplified, instead of exposing the storage details (identifiers, saved state changes, etc.) the interface could be simply `log`, `snapshot`, and `restore`